apps/internal/base/base.go

```diff
@@ -345,6 +349,26 @@ func (b Client) AcquireTokenSilent(ctx context.Context, silent AcquireTokenSilen
 	if silent.Claims == "" {
 		ar, err = AuthResultFromStorage(storageTokenResponse)
 		if err == nil {
+			if shouldRefresh(storageTokenResponse.AccessToken.RefreshOn.T) {
+				if tr, er := b.Token.Credential(ctx, authParams, silent.Credential); er == nil {
+					return b.AuthResultFromToken(ctx, authParams, tr)
+				} else {
+					var callErr *errors.CallErr
+					if errors.As(er, &callErr) {
+						switch callErr.Resp.StatusCode {
+						case http.StatusRequestTimeout, // 408
+							http.StatusTooManyRequests,     // 429
+							http.StatusInternalServerError, // 500
+							http.StatusBadGateway,          // 502
+							http.StatusServiceUnavailable,  // 503
+							http.StatusGatewayTimeout:      // 504
+						default:
+							// return empty token for non handable error
+							return AuthResult{}, er
```

Comment:
We should never return empty token. Instead, we should throw the error associated with the failure.

Reply:
This is the idiomatic equivalent to throwing--return the error and the zero value of any other types the function returns. Callers are responsible for checking whether the returned error is nil and if it isn't, they should discard anything else the function returned.

I don't understand why we'd throw here though. We could return the valid token from the cache. The fact that a refresh failed is interesting only when the application really needs a new token.

Reply:
Yes, sorry, my bad for forgetting the golang pattern. 

With regards to you question:

1. We don't want to keep retrying getting a token in all cases. For example if the client secret is bad, we shouldn't silently retry that. It would generate a lot of extra traffic for eSTS and the SDK's responsibility is to minimize that. refresh_in triggers only eSTS has a problem and they signal this by sending those status codes 
2. refresh_in is only issued by eSTS for some tenants where long lived tokens are enabled. So the token lifetime is much longer (~24h for app tokens) anyway, it won't cause a problem

Reply:
But this behavior will cause the application to fail when it could have succeeded given the cached token. Note that as written, the application gets an error when the authority doesn't respond, so the cache no longer makes the client resilient against transient network outages. There are other ways to avoid hammering the authority; we need backoff and synchronization here in any case. It will be rare for a client to get this far with an invalid secret, because in most usage the cache is in memory and begins empty, but we could handle that case by its error code.